README.markdown
# Tomcat extensions for Capistrano


## Usage

In your Rails app's deployment recipe (`config/deploy.rb`):

    require 'capistrano/tomcat'

Then, once you have built your WAR into `artifacts/appname.war`,
you can use the Capistrano deployment as you would for Passenger.

    $ rake package
    $ cap deploy:setup
    $ cap deploy

## Configuration

The default configuration uploads artifacts/yourapp.war to your app
servers and starts Tomcat on port 8080.

### Ports
* `tomcat_base_port`: controls the other ports as well. Defaults to 8005.
* `tomcat_ajp_port`: defaults to 8009 (`tomcat_base_port` + 4)
* `tomcat_http_port`: defaults to 8080 (`tomcat_base_port` + 75)
* `tomcat_https_port`: defaults to 8443 (`tomcat_base_port` + 438)

### Directories and files
* `catalina_home`: location of the base Tomcat installation, defaults to
  `/usr/share/tomcat6`
* `catalina_executable`: location of the Tomcat executable, defaults to `/usr/sbin/tomcat6`.  In a standard Tomcat installation this would be `${CATALINA_HOME}/bin/catalina.sh` or `${CATALINA_HOME}/bin/catalina.bat`

For the Tomcat installed with **Red Hat Enterprise Linux** packages,
the following settings are appropriate:

    cset :catalina_home '/usr/share/tomcat6'
    cset :catalina_executable, '/usr/sbin/tomcat6'

### Additional pieces
If you need Tomcat to run with additional environment variables, you can use the `tomcat_runtime_env` setting. For example:

    cset :tomcat_runtime_env, {'MYAPP_DB_HOSTNAME' => 'db.example.com'}

If you are using the Java keystore for SSL, you can set the password
with `keystore_password`.

### Unpacking the WAR

Some tasks and applications rely on the standard Rails files being
available on the filesystem. If you need to explode the WAR before
Tomcat gets to it, you can include the following lines in your recipe:

    after "deploy:update_code", "tomcat:unpack_war"
    after "tomcat:unpack_war", "tomcat:finalize_unpacked_war"



HISTORY.txt
v0.0.6
======
 * Compatible with Bundler 1.1 (thanks Jennifer Smith)

v0.0.5
======
 * Any servers with the 'worker' role will get a copy of the code, too.
   (thanks Eldon Alameda)

v0.0.4
======
 * Added targets for unpacking WARs
 * More useful README
 * Configurable `catalina_executable` for non-RedHat machines
   (thanks Marc Huffnagle)
 * HTTPS actually defaults to port 8443 like it was supposed to.
   (thanks Marc Huffnagle)

v0.0.3
======
 * Allows a deployment to add additional environment variables to the
   Tomcat runtime

v0.0.2
======
 * Deploys to a machine with RedHat Enterprise Linux and the `tomcat6` package.

v0.0.1
======
 * Initial release.
 * Actually a broken pre-release version, whoops.



lib/capistrano/tomcat.rb
require 'capistrano'

configuration = Capistrano::Configuration.respond_to?(:instance) ?
  Capistrano::Configuration.instance(:must_exist) :
  Capistrano.configuration(:must_exist)

configuration.load do
  set(:war_filename) { "#{application}.war" }
  _cset :tomcat_base_port, 8005
  _cset(:tomcat_ajp_port) { tomcat_base_port + 4 } #8009
  _cset(:tomcat_http_port) { tomcat_base_port + 75 } #8080
  _cset(:tomcat_https_port) { tomcat_base_port + 438 } #8443
  _cset :keystore_password, nil
  _cset :catalina_home, '/usr/share/tomcat6'
  _cset :catalina_executable, '/usr/sbin/tomcat6'
  _cset(:tomcat_runtime_env, {})

  namespace :deploy do
    task :update_code, :roles => [:app, :worker] do
      on_rollback { run "rm -rf -- \"#{release_path}\"" }

      create_tomcat_directory_structure_under release_path
      source_war_path = "artifacts/#{war_filename}"
      target_war_path = File.join(release_path, 'webapps', war_filename)
      top.upload source_war_path, target_war_path
    end

    task :finalize_update do
      # no finalization necessary with a WAR deployment
      # although it might be if you unpack the war with tomcat:unpack_war
    end

    task :start, :roles => :app do
      run_tomcat_command 'start'
    end
    task :stop, :roles => :app do
      run_tomcat_command 'stop'
    end
    task :restart, :roles => :app do
      run_tomcat_command 'stop'
      run_tomcat_command 'start'
    end

    def run_tomcat_command cmd
      log = File.join(shared_path, "log", "deploy.log")
      tmpdir = File.join(current_path, "temp")
      base_env = {
        "CATALINA_HOME" => catalina_home,
        "CATALINA_BASE" => current_path,
        "CATALINA_TMPDIR" => tmpdir
      }
      run "#{catalina_executable} #{cmd} #{log}", :env => base_env.merge(tomcat_runtime_env)
    end
  end

  def create_tomcat_directory_structure_under base
    dirs = [release_path,
      File.join(release_path, 'webapps'),
      File.join(release_path, 'conf'),
      File.join(release_path, 'work'),
      File.join(release_path, 'temp')
    ]
    run <<-SH
      #{try_sudo} mkdir -p #{dirs.join ' '} && \
      #{try_sudo} cp -r \"#{catalina_home}/conf\"/* \"#{release_path}/conf/\" && \
      #{try_sudo} ln -s #{File.join shared_path, "log"} #{File.join release_path, "logs"}
    SH

    upload_templated_config 'server.xml', :remote_base => base
    upload_templated_config 'web.xml', :remote_base => base
  end

  def upload_templated_config filename, options
    base = options[:remote_base] or raise ArgumentError "no base specified"

    destination_path = File.join(base, 'conf', filename)
    template_filename = File.join(File.dirname(__FILE__), "tomcat", "templates", "#{filename}.erb")
    template = File.read(template_filename)
    result = ERB.new(template).result(binding)
    put result, destination_path, :mode => 0644
  end

  namespace :tomcat do
    desc "Unpack a WAR for tools that need the individual files"
    task :unpack_war, :roles => [:app, :worker] do
      target_webapps_path = File.join(release_path, 'webapps')
      target_war_path = File.join(target_webapps_path, war_filename)
      target_unpacked_war_path = unpacked_war_path_under release_path

      on_rollback { run "rm -rf -- \"#{target_unpacked_war_path}\"" }
      run %{unzip -q "#{target_war_path}" -d "#{target_unpacked_war_path}"}
    end

    desc "Touch up an unpacked WAR to look like a regular Rails app"
    task :finalize_unpacked_war, :roles => [:app, :worker] do
      unpacked_war_path = unpacked_war_path_under release_path
      app_path = File.join unpacked_war_path, 'WEB-INF'
      # code duplicated from capistrano's deploy recipe:
      # (lib/capistrano/recipes/deploy.rb)
      run <<-CMD
      rm -rf #{app_path}/log #{app_path}/public/system #{app_path}/tmp/pids &&
      mkdir -p #{app_path}/public &&
      mkdir -p #{app_path}/tmp &&
      ln -s #{shared_path}/log #{app_path}/log &&
      ln -s #{shared_path}/system #{app_path}/public/system &&
      ln -s #{shared_path}/pids #{app_path}/tmp/pids
      CMD
    end

    def unpacked_war_path_under(release_path)
      webapps_path = File.join(release_path, 'webapps')
      war_path = File.join(webapps_path, war_filename)
      unpacked_war_path = File.join(webapps_path, File.basename(war_filename, '.war'))
    end
  end
end

